src/tools/kata-ctl/src/arch/mod.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{Result};
+
+#[cfg(target_arch = "aarch64")]
+pub mod aarch64;
+
+#[cfg(target_arch = "powerpc64le")]
+pub mod powerpc64le;
+
+#[cfg(target_arch = "s390x")]
+pub mod s390x;
+
+#[cfg(target_arch = "x86_64")]
+pub mod x86_64;
+
+pub fn check(global_args: clap::ArgMatches) -> Result<()> {
+    #[cfg(target_arch = "aarch64")]
+    let result = aarch64::check();
+
+    #[cfg(target_arch = "powerpc64le")]
+    let result = powerpc64le::check();
+
+    #[cfg(target_arch = "s390x")]
+    let result = s390x::check();
+
+    #[cfg(target_arch = "x86_64")]
+    let result = x86_64::check(global_args);
+
+    #[cfg(not(any(
+        target_arch = "aarch64",
+        target_arch = "powerpc64le",
+        target_arch = "s390x",
+        target_arch = "x86_64"
+    )))]
+    panic!("unknown architecture");
```

Comment:
For common functions that all architecture needs to provide and share the same signature, a nicer way is exposing these common functions directly from the arch-specific module with the right `cfg` guard, for example:
```
#[cfg(target_arch = "x86_64")]
pub use x86_64::check;

#[cfg(target_arch = "aarch64")]
pub use aarch64::check;

...
```
In this way, the actual interface function of the crate (e.g. `arch::check()`) will be better abstracted and simpler. This will also report error at compile time for unsupported architectures (instead of a panic at runtime).

Reply:
Hi @likebreath - I'm confused. Firstly, that panic line was replaced on commit 7c9f9a5a1dae3f9c2c16748b6d5af9fc134abe65 so I'm not sure why/how you were seeing that. Secondly, we're already using cfg guards. The code I think you might be referring to (now) is:

```rust
#[cfg(not(any(
    target_arch = "aarch64",
    target_arch = "powerpc64le",
    target_arch = "s390x",
    target_arch = "x86_64"
)))]
compile_error!("unknown architecture");
```

... and I think that does exactly what we want: it will fail to compile on unknown platforms.

Reply:
Ah - I see what you're saying about the cfg guard now. Must be time for :coffee: :smile: